### DISCUSSION TITLE
Insights on error when moving from c99 to c17 in scikit-learn meson.build?

### DISCUSSION THREAD

**Discussion by lesteve** [pyodide/pyodide#4762] [Q&A]
I seem to be able to do `emcc -std=c17 test.c` with emscripten 3.1.45 so not sure why meson is complaining about this:
```
../meson.build:1:0: ERROR: None of values ['c17'] are supported by the C compiler. Possible values are ['none', 'c89', 'c99', 'c11', 'gnu89', 'gnu99', 'gnu11']
```

Could it be an issue in pywasmcross.py somehow?

See https://github.com/scikit-learn/scikit-learn/issues/29013 for more details about how it manifests itself in scikit-learn.

**Comment by rgommers** [comment]
This is almost certainly a problem in Meson itself. The supported C standard versions for Emscripten are inherited from Clang, but the Emscripten version numbers are much lower than the Clang ones. Clang 6.0 is the minimum version for C17 support (https://github.com/mesonbuild/meson/blob/cfd57180eef9036c7167c5682b9f3055a540fccc/mesonbuild/compilers/c.py#L112), but Emscripten is only at version 3.x so it'll fail that check if it's not special-cased.

I can open a PR for that tonight probably. Just need to figure out what the correct minimum version is for declaring C17 (and C18/C20) support.

**Reply by rgommers** [reply]
@lesteve I'm traveling and have no install at hand. It looks like there is no public list of Emscripten <--> Clang versions, but it can be obtained from `emcc -v` and `./emsdk list --old` (says https://emscripten.org/docs/tools_reference/emsdk.html#how-do-i-install-and-activate-old-emscripten-sdks-and-tools). If that works, would you be able to check the versions of Emscripten that correspond to Clang 6.0.0, 8.0.0, 9.0.0, 18.0.0? 

This should fix the problem I believe (again, I can't check right now): https://github.com/rgommers/meson/tree/emscripten-c17

**Reply by lesteve** [reply]
- clang >= 18.0.0 corresponds to emscripten >= 3.0.0
- for the other clang >= 6.0.0, clang >= 8.0.0 and clang >= 9.0.0, it's not clear or they correspond to older versions that what is available through emsdk? Maybe @hoodmane or @ryanking13, you know a better way to have the relationship between Emscripten and Clang versions than downloading a particular version through emsdk and doing `emcc -v`?

Those were the particular clang versions that you care about If I understand your commit https://github.com/mesonbuild/meson/compare/master...rgommers:meson:emscripten-c17 
```python
_C17_VERSION = '>=3.0.0'  # '>=6.0.0'
_C18_VERSION = '>=3.0.0'  # '>=8.0.0'
_C2X_VERSION = '>=3.0.0'  # '>=9.0.0'
_C23_VERSION = '>=3.0.0'  # '>=18.0.0'
```

Here is a table for a few versions (1.38.33 is the older version available through emsdk):
```
| emsdk version |         clang version |
|---------------+-----------------------|
|    1.38.33-34 | error getting version |
|       1.38.35 |                 9.0.0 |
|        1.39.0 |                10.0.0 |
|         2.0.0 |                12.0.0 |
|        2.0.34 |                14.0.0 |
|        3.1.45 |                18.0.0 |
```

I got them through a hacky script like this:
```bash
set -e

func() {
    version=$1
    echo installing $version
    ./emsdk install $version > /dev/null 2>&1
    echo done installing
    ./emsdk activate $version > /dev/null 2>&1
    source emsdk_env.sh > /dev/null 2>&1
    echo emcc version $version
    echo clang version $(emcc -v 2>&1 | grep 'clang version' || echo error getting version)
}

# for version in 1.38.40 1.39.0 2.0.0 3.0.0; do
# for version in 1.38.34 1.38.35 1.38.36 1.38.37 1.38.38 1.38.39; do
#     func $version
# done

func $1

```

**Reply by rgommers** [reply]
Great, thanks @lesteve. I opened https://github.com/mesonbuild/meson/pull/13221 and added you as a co-author.

**Reply by QuLogic** [reply]
> clang >= 18.0.0 corresponds to emscripten >= 3.0.0

This doesn't seem right:
```
$ emcc -v
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.0.0 (3fd52e107187b8a169bb04a02b9f982c8a075205)
clang version 14.0.0 (https://github.com/llvm/llvm-project 4348cd42c385e71b63e5da7e492172cff6a79d7b)
```

**Reply by lesteve** [reply]
Indeed I double-checked and the correct version for clang 18.0.0 is actually 3.1.45.

```
❯ emcc -v
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.45 (ef3e4e3b044de98e1811546e0bc605c65d3412f4)
clang version 18.0.0 (https://github.com/llvm/llvm-project d1e685df45dc5944b43d2547d0138cd4a3ee4efe)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /home/lesteve/dev/emsdk/upstream/bin
```

**Reply by lesteve** [reply]
I have opened https://github.com/mesonbuild/meson/pull/13764 to fix this
